### DISCUSSION TITLE
Zero-or-one to zero-or-one relations

### DISCUSSION THREAD

**Discussion by giacomo-cc** [supabase/supabase#12914] [Questions] [answered]
Hi, we're using supabase postgrest wrapper in production but I'm facing some errors with schema modelling right now.
I'd like to provide a Zero-or-one to zero-or-one relation between two tables, but with no foreign key between the two the supabase postgrest fails. What's the best way to model it? Thanks

**Comment by giacomo-cc** [comment, marked as answer]
I found a way to do so creating a function on the database to establish the relation, here's my model:

```sql
CREATE TABLE test_contracts (
	pubkey varchar NOT NULL,
	value varchar NOT NULL,
	CONSTRAINT test_contracts_pk PRIMARY KEY (pubkey)
);

CREATE TABLE test_metadata (
	pubkey varchar NOT NULL,
	metadata varchar NULL,
	CONSTRAINT test_metadata_pk PRIMARY KEY (pubkey)
);

create function test_metadata_fn(test_contracts) returns setof test_metadata rows 1 as $$
  select * from test_metadata where pubkey = $1.pubkey
$$ stable language sql;
```

and then with postgrest:

the query that doesn't work without the `function` because of the missing FK:
```ts
const res = await new PostgrestClient<Database>(url).from("test_contracts").select(`*,test_metadata(*)`);
```
the query that finally works:
```ts
const res = await new PostgrestClient<Database>(url).from("test_contracts").select(`*,test_metadata_fn(*)`);
```

I've found the tip on the postgrest docs, [here](https://postgrest.org/en/stable/api.html#computed-relationships)

I'm wondering if my implementation is the best I can get
cc @steve-chavez @GaryAustin1

**Reply by steve-chavez** [reply]
@thegreat-vanderlinde Yes, if you don't have foreign keys between tables computed relationships are the way to go.

Note that you can also use it on SQL like:

```sql
SELECT *, test_metadata_fn(x) FROM test_contracts x;
```

**Reply by giacomo-cc** [reply]
Great thank you! 🙏🏼

Can you also provide some help on the `Database` typing?

This is the `Database` type I'm using with `new PostgrestClient<Database>(url);`:
```ts
export type Database = {
	public: {
		Tables: {
			[CONTRACTS_TABLE_NAME]: Table<ContractSchema>;
			[CONTRACTS_METADATA_TABLE_NAME]: Table<MetadataSchema>;
		};
		Views: {};
		Functions: {
			[CONTRACTS_METADATA_FN_NAME]: {
				Args: { name_param: string }; // ???
				Returns: Database['public']['Tables']["contracts_metadata"]; // ???
			};
		};
	};
};
```

I'm trying multiple options, taking inspiration from [postgrest-js/tests](https://github.com/supabase/postgrest-js/blob/master/test/types.ts) but I can't find a way to get my types correctly injected.

**Reply by GaryAustin1** [reply]
@thegreat-vanderlinde You should not add on radically different questions from the original and instead start a new discussion.

You might get better response on Typescript / Supabase typing questions on the Supabase Discord, or maybe even find your answer there.

**Reply by giacomo-cc** [reply]
@GaryAustin1 my question was related to the sql function used to solve the original question. I'll check on Discord for similar situations, thanks for the input
